federation/pkg/federatedtypes/hpa.go

```diff
@@ -936,3 +944,60 @@ func (a *HpaAdapter) minReplicasIncreasable(obj pkgruntime.Object) bool {
 	}
 	return false
 }
+
+// updateClusterListOnTargetObject passes the necessary info to the target object,
+// so that the corresponding controller can act on that.
+// This is used because if an hpa is active on a federated object it is supposed
+// to control the replicas and presence/absence of target object from federated clusters.
+func (a *HpaAdapter) updateClusterListOnTargetObject(fedHpa *autoscalingv1.HorizontalPodAutoscaler, scheduleStatus map[string]*replicaNums) error {
+	if len(fedHpa.Spec.ScaleTargetRef.Kind) <= 0 || len(fedHpa.Spec.ScaleTargetRef.Name) <= 0 {
+		// nothing to do
+		return nil
+	}
+
+	names := []string{}
+	for clusterName, replicas := range scheduleStatus {
+		if replicas != nil {
+			names = append(names, clusterName)
+		}
+	}
+	clusterNames := hpautil.ClusterNames{Names: names}
+	qualifiedKind := extensionsinternal.Kind(fedHpa.Spec.ScaleTargetRef.Kind)
+	targetObj, err := getRuntimeObjectForKind(a.client, qualifiedKind, fedHpa.Namespace, fedHpa.Spec.ScaleTargetRef.Name)
+	if errors.IsNotFound(err) {
+		// Nothing to do; the target object does not exist in federation.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	updatedObj := hpautil.SetHpaTargetClusterList(targetObj, clusterNames)
+	_, err = updateRuntimeObjectForKind(a.client, qualifiedKind, fedHpa.Namespace, updatedObj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getRuntimeObjectForKind(c federationclientset.Interface, kind schema.GroupKind, ns, name string) (pkgruntime.Object, error) {
+	switch kind {
+	case extensionsinternal.Kind("ReplicaSet"):
+		return c.ExtensionsV1beta1().ReplicaSets(ns).Get(name, metav1.GetOptions{})
+	case extensionsinternal.Kind("Deployment"):
+		return c.ExtensionsV1beta1().Deployments(ns).Get(name, metav1.GetOptions{})
+	default:
+		return nil, fmt.Errorf("Unsupported federated kind targeted by hpa: %v", kind)
+	}
+}
+
+func updateRuntimeObjectForKind(c federationclientset.Interface, kind schema.GroupKind, ns string, obj pkgruntime.Object) (pkgruntime.Object, error) {
```

Comment:
Actually, I did look to do something similar what you suggest, but to my surprise found that, at almost all places, where ever the kind or the resource name is initialised, its used as a string directly.
For example [federation-apiserver default resource config](https://github.com/kubernetes/kubernetes/blob/master/federation/cmd/federation-apiserver/app/server.go#L267) (which is done quite similar in kube-apiserver) or specifying a kind, [as in deployment controller ref kind](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L417) or [the replicaset controller kind itself](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L62).
I did go ahead to use the string as in the code reviewed, thus.

Additionally, in k8s api, the hpa targetref object is not interpreted (as is done here) within the hpa controllers purview. The scales subresource is used instead. So what ever target name and kind comes up in the hpa spec, the scales api is called using the same. Only particular types support this subresource, thus the api call will fail, if called on a wrong type or kind (which probably was wrong in the hpa spec, specified by the user), [ref](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L384). 
I did use only "Deployment" and "ReplicaSet" in the kind check here, because those are the only two kinds which are available against federation API.
I am not sure, how else can I handle it. On the other hand, I have limited understanding of the API machinery; @nikhiljindal would you kindly be able to help a bit on this, thanks!